### PR TITLE
Add a scope for launching background work in tests

### PR DESCRIPTION
**Pull request by dkhalanskyjb**
Fixes https://github.com/Kotlin/kotlinx.coroutines/issues/3287

### CODE REVIEW

File: kotlinx-coroutines-test/common/test/TestScopeTest.kt [hunk 1]
```diff
@@ -169,6 +169,167 @@ class TestScopeTest {
...
+        delay(10)
+        assertEquals(1, i)
+        assertEquals(1, j)
+    }
+
+    /**
+     * Tests that the background work gets cancelled after the test body finishes.
+     */
+    @Test
+    @NoNative
```
Comment (dkhalanskyjb) [inline kotlinx-coroutines-test/common/test/TestScopeTest.kt:198]: This is a draft until we can make the new tests run on Native.

**Comment by dkhalanskyjb** [comment]
Reported the compiler bug here: https://youtrack.jetbrains.com/issue/KT-52985/Native-a-function-with-type-T-returned-a-kotlinUnit-instead-of-null

File: kotlinx-coroutines-test/common/src/TestScope.kt [hunk 2]
```diff
@@ -46,6 +46,23 @@ public sealed interface TestScope : CoroutineScope {
...
+    /**
+     * A scope for background work.
+     *
+     * This scope is automatically cancelled when the test finishes.
+     * Additionally, while the coroutines in this scope are run as usual when
+     * using [advanceTimeBy] and [runCurrent], [advanceUntilIdle] will stop advancing the virtual time
+     * once only the coroutines in this scope are left unprocessed.
+     *
+     * Failures in coroutines in this scope do not terminate the test.
+     * Instead, they are reported at the end of the test.
```
Comment (qwwdfsad) [inline kotlinx-coroutines-test/common/src/TestScope.kt:59]: It worth mentioning that this scope is not bound by parent-child relationship with test scope itself, so any failure in the test itself won't affect this scope

File: kotlinx-coroutines-test/common/src/TestScope.kt [hunk 3]
```diff
@@ -170,6 +187,9 @@ internal class TestScopeImpl(context: CoroutineContext) :
     private val uncaughtExceptions = mutableListOf<Throwable>()
     private val lock = SynchronizedObject()
 
+    override val backgroundWorkScope: CoroutineScope =
+        CoroutineScope(coroutineContext + SupervisorJob() + BackgroundWork)
```
[the remark this answers is not in the document]
Comment (dkhalanskyjb) [reply kotlinx-coroutines-test/common/src/TestScope.kt:191]: This reveals that we need to improve the name/framing.

The purpose of this scope is not the fuzzy notion of background work, but, specifically, for coroutines that, from the perspective of the test, are always there. They are only cancelled so that resources are not leaked; semantically, they keep running for the whole duration of the test. I would even expect such background work to be launched outside of the test definition, in fields of the test class etc. So, if something significant happens on cancellation of a background task, using this scope is not the proper way to check that.

[the remark this answers is not in the document]
Comment (qwwdfsad) [reply kotlinx-coroutines-test/common/src/TestScope.kt:191]: Even with a better name, the problem will still exist. It would be nice to report _any_ exceptions that happen within the scope

[the remark this answers is not in the document]
Comment (dkhalanskyjb) [reply kotlinx-coroutines-test/common/src/TestScope.kt:191]: Not sure about that.

Given the more clear definition of background work, we see that, typically, the background work executes in places that suffer from the same problem, only in production. For example, `viewModelScope` uses `SupervisorJob`, and something crashing in `GlobalScope` is notoriously elusive.

So, let's consider the possibilities.
* The background code never crashes silently in production. Ok, fair enough, nothing to discuss.
* It may crash silently in production, but in a controllable manner that the foreground code expects. Then we would just complicate the tests with an unneeded error-handling ceremony if we forced the error to surface.
* It may crash silently in production, and this is unexpected and should be tested for. Well, why is that not a separate test then?

File: kotlinx-coroutines-test/common/src/TestScope.kt [hunk 4]
```diff
@@ -46,6 +47,43 @@ public sealed interface TestScope : CoroutineScope {
      */
     @ExperimentalCoroutinesApi
     public val testScheduler: TestCoroutineScheduler
+
+    /**
+     * A scope for background work.
+     *
+     * This scope is automatically cancelled when the test finishes.
```
Comment (whyoleg) [inline kotlinx-coroutines-test/common/src/TestScope.kt:54]: will `runTest` await until `backgroundScope` finished after cancelling? Or it will just cancel it and complete test leaving `backgroundScope` cancelling in background?

Comment (dkhalanskyjb) [reply kotlinx-coroutines-test/common/src/TestScope.kt:54]: It will await the background scope. There's no "background", that is, no separate thread for this.

File: kotlinx-coroutines-test/common/test/TestScopeTest.kt [hunk 5]
```diff
@@ -169,6 +169,164 @@ class TestScopeTest {
         }
     }
 
+    /** Tests that the background work is being run at all. */
+    @Test
+    fun testBackgroundWorkBeingRun(): TestResult = runTest {
```
[the remark this answers is not in the document]
Comment (dkhalanskyjb) [reply kotlinx-coroutines-test/common/test/TestScopeTest.kt:176]: Provided a test in that style.

File: kotlinx-coroutines-test/common/src/TestScope.kt (hunk 2 above)
Comment (dkhalanskyjb) [reply kotlinx-coroutines-test/common/src/TestScope.kt:60]: Done.

File: kotlinx-coroutines-test/common/src/TestScope.kt (hunk 3 above)
[the remark this answers is not in the document]
Comment (dkhalanskyjb) [reply kotlinx-coroutines-test/common/src/TestScope.kt]: After an internal discussion, we arrived at this solution: do keep `SupervisorJob`, because it allows the background tasks to fail independently of the test code and one another, which would, in our opinion, lead to the least surprising behavior, but *do* report all the exceptions that it is possible for us to observe. For now, we will assume that the actual production code does not rely on `async` and `produce` not surfacing the errors until queried. If this helps surface some errors, great. If it turns out that people *are* using the fact that exceptions are not reported, we will learn about this with actual use cases. In general, we follow the rule of thumb that exceptions are probably something that's not to be unhandled.

File: kotlinx-coroutines-test/common/src/TestScope.kt [hunk 6]
```diff
@@ -46,6 +46,23 @@ public sealed interface TestScope : CoroutineScope {
...
+     * once only the coroutines in this scope are left unprocessed.
+     *
+     * Failures in coroutines in this scope do not terminate the test.
+     * Instead, they are reported at the end of the test.
+     *
+     * A typical use case for this scope is to launch tasks that would outlive the tested code in
+     * the production environment.
+     */
+    @ExperimentalCoroutinesApi
+    public val backgroundWorkScope: CoroutineScope
```
[the remark this answers is not in the document]
Comment (dkhalanskyjb) [reply kotlinx-coroutines-test/common/src/TestScope.kt]: Sure. I assumed that even `backgroundScope` would be too long, so people would do something like
```kotlin
backgroundScope.apply {
    launch { ... }
    launch { ... }
}
```
or pass the scope somewhere. Then I couldn't think of an even shorter name, so I decided to go all-out. However, if people are already using the name `backgroundScope`, it's a great way to tap into an existing convention.

File: kotlinx-coroutines-test/common/src/TestBuilders.kt [hunk 7]
```diff
@@ -216,16 +220,27 @@ internal suspend fun <T: AbstractCoroutine<Unit>> runTestCoroutine(
...
-        select<Unit> {
-            coroutine.onJoin {
-                completed = true
-            }
-            scheduler.onDispatchEvent {
-                // we received knowledge that `scheduler` observed a dispatch event, so we reset the timeout
+        // in case progress depends on some background work, we need to keep spinning it.
+        val backgroundWorkRunner = launch(CoroutineName("background work runner")) {
+            while (true) {
+                scheduler.tryRunNextTaskUnless { !isActive }
```
Comment (qwwdfsad) [inline kotlinx-coroutines-test/common/src/TestBuilders.kt:226]: Is my understanding of this machinery correct?

* We cannot run background jobs in the previous `advanceUntilIdle` because otherwise, we may be stuck in an endless loop of background work
* `yield` is required to get the chance to `select` to complete and switch `isActive` trigger
* We cannot avoid this `launch` even though we get notified in `onDispatchEvent` because of p.1

Comment (qwwdfsad) [reply kotlinx-coroutines-test/common/src/TestBuilders.kt:226]: I've tried to figure out the way to avoid ceremonies with extra `launch`, but it doesn't seem to be that easy.

The machinery itself is already quite complex and it would be nice if you could drop a comment with a rationale behind it

File: kotlinx-coroutines-test/common/src/TestBuilders.kt [hunk 8]
```diff
@@ -216,16 +220,27 @@ internal suspend fun <T: AbstractCoroutine<Unit>> runTestCoroutine(
...
-            onTimeout(dispatchTimeoutMs) {
-                handleTimeout(coroutine, dispatchTimeoutMs, tryGetCompletionCause, cleanup)
+        }
+        try {
+            select<Unit> {
+                coroutine.onJoin {
+                    completed = true
+                }
+                scheduler.onDispatchEvent {
+                    // we received knowledge that `scheduler` observed a dispatch event, so we reset the timeout
```
Comment (qwwdfsad) [inline kotlinx-coroutines-test/common/src/TestBuilders.kt:236]: also worth noting that it doesn't receive events from background jobs

File: kotlinx-coroutines-test/common/src/internal/ReportingSupervisorJob.kt [hunk 9]
```diff
@@ -0,0 +1,16 @@
...
+
+import kotlinx.coroutines.*
+
+/**
+ * A variant of [SupervisorJob] that additionally notifies about child failures via a callback.
+ */
+@Suppress("INVISIBLE_MEMBER", "INVISIBLE_REFERENCE", "CANNOT_OVERRIDE_INVISIBLE_MEMBER")
+internal class ReportingSupervisorJob(parent: Job? = null, val onChildCancellation: (Throwable) -> Unit) :
+    JobImpl(parent) {
+    override fun childCancelled(cause: Throwable): Boolean = onChildCancellation(cause).let { false }
```
Comment (qwwdfsad) [inline kotlinx-coroutines-test/common/src/internal/ReportingSupervisorJob.kt:15]: It's potato-potato of course, but personally I find multi-liners here much more straightforward:
```
onChildCancellation(cause)
return false
```

File: kotlinx-coroutines-test/common/src/TestScope.kt [hunk 10]
```diff
@@ -170,14 +208,56 @@ internal class TestScopeImpl(context: CoroutineContext) :
     private val uncaughtExceptions = mutableListOf<Throwable>()
     private val lock = SynchronizedObject()
 
+    override val backgroundScope: CoroutineScope =
+        CoroutineScope(coroutineContext + BackgroundWork + ReportingSupervisorJob {
+            if (it !is CancellationException) reportException(it)
```
Comment (qwwdfsad) [inline kotlinx-coroutines-test/common/src/TestScope.kt:213]: `reportException` may throw an exception in edge cases. Job's machinery is unlikely to be ready for that and it can lead to very weird deadlocks or inconsistent states.

I suggest doing something else, for example reporting it to global EH

File: kotlinx-coroutines-test/common/src/TestScope.kt [hunk 11]
```diff
@@ -170,14 +208,56 @@ internal class TestScopeImpl(context: CoroutineContext) :
...
+     * ```
+     */
+    // Only call this in a `synchronized(lock)` block.
+    @Suppress("INVISIBLE_MEMBER")
+    private fun collectUncaughtExceptions(): List<Throwable> {
+        val exceptions = LinkedHashSet(uncaughtExceptions)
+        val unwrappedExceptions = LinkedHashSet(uncaughtExceptions.map { unwrap(it) })
+        for (child in backgroundScope.coroutineContext.job.children) {
+            val exception = try {
+                child.getCancellationException()
```
Comment (qwwdfsad) [inline kotlinx-coroutines-test/common/src/TestScope.kt:243]: You can probably guard it by `if (!child.isCancelled) continue` to save some time

File: kotlinx-coroutines-test/common/src/TestScope.kt [hunk 12]
```diff
@@ -170,14 +208,56 @@ internal class TestScopeImpl(context: CoroutineContext) :
...
+     *     throw TestException()
+     *   }
+     *   delay(1)
+     * }
+     * ```
+     */
+    // Only call this in a `synchronized(lock)` block.
+    @Suppress("INVISIBLE_MEMBER")
+    private fun collectUncaughtExceptions(): List<Throwable> {
+        val exceptions = LinkedHashSet(uncaughtExceptions)
```
Comment (qwwdfsad) [inline kotlinx-coroutines-test/common/src/TestScope.kt:239]: In `JobSupport` we use an identity hash set to workaround throwables that throw exceptions from their `hashCode`/`equals`. 
I don't think it's worth fixing here, but it's definitely worth to file an issue here and I'll escalate it to MPP stdlib

File: kotlinx-coroutines-test/common/src/TestScope.kt (hunk 10 above)
Comment (dkhalanskyjb) [reply kotlinx-coroutines-test/common/src/TestScope.kt:213]: Do you have good ideas for the `context` for the `handleCoroutineException` invocation?

File: kotlinx-coroutines-test/common/src/internal/ReportingSupervisorJob.kt (hunk 9 above)
Comment (dkhalanskyjb) [reply kotlinx-coroutines-test/common/src/internal/ReportingSupervisorJob.kt]: I like the one-liner approach because the exit points are self-evident. In the example you commented on, yes, this can go both ways. The code was modified and has more complex control flow, but the intent to always return `false` didn't change, and here the benefit, if there is any, should be more clear: the code clearly signifies that, no matter what happens, `false` should always be returned here. One doesn't have to scan the function for other possible exit points.

File: kotlinx-coroutines-test/common/src/TestScope.kt (hunk 11 above)
Comment (dkhalanskyjb) [reply kotlinx-coroutines-test/common/src/TestScope.kt]: Or I can just remove the code altogether.

File: kotlinx-coroutines-test/common/src/TestScope.kt [hunk 13]
```diff
@@ -218,6 +260,12 @@ internal class TestScopeImpl(context: CoroutineContext) :
             if (finished) {
                 throw throwable
             } else {
+                @Suppress("INVISIBLE_MEMBER")
+                for (existingThrowable in uncaughtExceptions) {
+                    // avoid reporting exceptions that already were reported.
+                    if (unwrap(throwable) == unwrap(existingThrowable))
```
[the remark this answers is not in the document]
Comment (dkhalanskyjb) [reply kotlinx-coroutines-test/common/src/TestScope.kt:266]: I looked at this for inspiration:
```kotlin
private fun Throwable.isSameExceptionAs(other: Throwable?): Boolean =
    other != null && unwrap(other) == unwrap(this)
```

**Review by qwwdfsad — APPROVED** [review]
